rfcs/0039-buffer-physical-layout.md

```diff
@@ -0,0 +1,640 @@
+    transformations in the TIR graph, which may or may not change the
+    dimensionality of the buffer.  The flattening of buffers to a
+    device-supported dimensionality would be handled identically to
+    any other layout transformation, rather than having an implicit
+    row-major traversal.
+
+  - Option 2: The `te.AXIS_SEPARATOR` is represented in the TIR graph,
+    and alters the behavior of the `StorageFlatten` pass.  There is no
+    `BufferTransform` node that represents the flattening of
+
+    In a TIR graph without any other modifications, this would
+    maintain the current behavior of the `StorageFlatten` pass, which
+    reduces the N-d buffer to a 1-d buffer by a row-major traversal.
+    In a TIR graph with some additional annotation to represent the
+    `M` axis separators, the N-d buffer could instead be reduced to a
+    `M+1`-d buffer.
+
+- What is appropriate terminology for size/shape/extent of physical
+  and logical buffers?
+
+  If Allocate/BufferStore/BufferLoad each hold a reference to the
+  buffer they act upon, then this becomes a somewhat irrelevant
+  question, as there is only one `BufferNode::shape`.
+
+  - I am partial to using "shape" both for the N-d parameters, and
+    have attempted to use it consistently through this RFC.
+  - "size" implies a 1-d buffer, which wouldn't be appropriate for
+    an N-d parameter.
+  - "extent" would be a reasonable name, but is currently used by
+    `tvm::RangeNode` to indicate a range of values that may start at
+    a non-zero value.  Since the indices for logical and physical
+    buffers both start at zero, using "extents" for the maximum
+    index would imply some offset.
+
+
+
+- How should loops over an array be handled when re-writing the shape?
+
+  To avoid memory latency issues, loops should iterate over an array
+  sequentially sequentially when possible.  Iteration that is defined
+  in terms of the logical layout may be inappropriate for the physical
+  layout.
+
+  Option 3 is preferred.
+
+  - Option 1: Do nothing, and always keep the same iteration order,
+    using the same iteration axes as defined in the compute
+    definition.
+
+    This would produce valid code, but not necessarily performant
+    code.  This can be a default behavior during development, to be
+    improved upon.
+
+  - Option 2: Automatically detect loops that are over the full extent
+    of an array in sequential order of the logical layout, and rewrite
+    to be in sequential order of the physical layout.
+
+    This would reduce the memory latency issues, but raises some
+    implementation questions.
+
+    - If a loop body references multiple tensors with different
+      physical layouts, which should define the loop iteration order?
+
+    - If a series of nested loops contains a `cache_read` or
+      `cache_write` stage, can these be recognized and reordered?
+
+  - Option 3: Expose the transformed axes to be used as part of a
```

Comment:
does this mean you'd write the schedule in terms of the physical layout? if so then many of my previous concerns about debuggability go away.

Reply:
It does, yes.  The `transform_layout` method will return an object representing the transformed tensor.  Later calls, such as `s.vectorize` or `s.reorder`, will accept the axes of the transformed tensor.